dbms/src/Interpreters/NullAwareSemiJoinHelper.cpp

```diff
@@ -63,26 +68,41 @@ void NASemiJoinResult<KIND, STRICTNESS>::fillRightColumns(MutableColumns & added
     }
     else if constexpr (STEP == NASemiJoinStep::NOT_NULL_KEY_CHECK_NULL_ROWS || STEP == NASemiJoinStep::NULL_KEY_CHECK_NULL_ROWS)
     {
-        for (size_t i = 0; i < max_pace; ++i)
+        size_t count = pace;
+        while (null_rows_pos < null_rows.size())
         {
-            if (null_rows_it == nullptr)
+            const auto & rows = null_rows[null_rows_pos];
+
+            while (columns_vec_pos < rows.materialized_columns_vec.size())
             {
-                if (null_rows_pos >= null_rows.size())
-                {
-                    step_end = true;
+                const auto & columns = rows.materialized_columns_vec[columns_vec_pos];
+
+                size_t insert_cnt = std::min(count, columns[0]->size() - columns_index);
```

Comment:
Don't insert if `insert_cnt` is 0?

Reply:
I change the logic here. Now `insert_cnt` should not be 0 as long as the `columns[0]->size()` is not 0.